File: typescript/src/data-structures/stack.ts
export abstract class Stack<T> {
  /** Returns the top of the stack. */
  abstract peek(): T;

  /** Adds a value to the top of the stack. */
  abstract push(elem: T): void;

  /** Removes and returns the top of the stack. */
  abstract pop(): T;

  /** Number of elements in the stack. */
  abstract size(): number;

  /** Clear all elements in the stack. */
  abstract clear(): void;

  isEmpty() {
    return this.size() === 0;
  }
}

export class ArrayStack<T> extends Stack<T> {
  protected elems: T[];

  constructor() {
    super();
    this.elems = [];
  }

  peek() {
    if (this.elems.length === 0) {
      throw new Error("Stack is empty");
    }
    return this.elems[this.elems.length - 1];
  }

  push(elem: T) {
    this.elems.push(elem);
  }

  pop() {
    if (this.elems.length === 0) {
      throw new Error("Stack is empty");
    }
    return this.elems.pop();
  }

  size() {
    return this.elems.length;
  }

  clear() {
    this.elems = [];
  }
}

export interface StackNode<T> {
  value: T;
  next: StackNode<T>;
}

export class LinkedStack<T> extends Stack<T> {
  protected top: StackNode<T>;
  protected _size: number;

  constructor() {
    super();
    this.top = undefined;
    this._size = 0;
  }

  peek() {
    if (!this.top) {
      throw new Error("Stack is empty");
    }
    return this.top.value;
  }

  push(value: T) {
    const node: StackNode<T> = {
      value,
      next: this.top
    };
    this.top = node;
    this._size++;
  }

  pop() {
    if (!this.top) {
      throw new Error("Stack is empty");
    }
    const { value } = this.top;
    this.top = this.top.next;
    this._size--;

    return value;
  }

  size() {
    return this._size;
  }

  clear() {
    this.top = undefined;
    this._size = 0;
  }
}

File: ruby/src/data-structures/stack.rb
# frozen_string_literal: true

EMPTY_STACK_EXCEPTION = 'Stack is empty'

module DataStructures
  class ArrayStack
    def initialize
      @elems = []
    end

    def peek
      return raise EMPTY_STACK_EXCEPTION if @elems.length.zero?

      @elems[@elems.length - 1]
    end

    def push(value)
      @elems.push(value)
    end

    def pop
      return raise EMPTY_STACK_EXCEPTION if @elems.length.zero?

      @elems.pop
    end

    def size
      @elems.length
    end

    def clear
      @elems = []
    end

    def to_array
      @elems.clone
    end

    def empty?
      @elems.length.zero?
    end
  end

  class LinkedStack
    attr_reader :size

    class StackNode
      attr_accessor :value, :next

      def initialize(value, next_node)
        @value = value
        @next = next_node
      end
    end

    def initialize
      @top = nil
      @size = 0
    end

    def peek
      return raise EMPTY_STACK_EXCEPTION if @top.nil?

      @top.value
    end

    def push(value)
      node = StackNode.new(value, @top)
      @top = node
      @size += 1
    end

    def pop
      return raise EMPTY_STACK_EXCEPTION if @top.nil?

      value = @top.value
      @top = @top.next
      @size -= 1

      value
    end

    def clear
      @top = nil
      @size = 0
    end

    def to_array
      current = @top
      result = []
      until current.nil?
        result.push(current.value)
        current = current.next
      end
      result.reverse
    end

    def empty?
      @top.nil?
    end
  end
end
